pkg/tinydtls/contrib/sock_dtls.c

```diff
@@ -0,0 +1,474 @@
+            return -ETIMEDOUT;
+        }
+    }
+    return 0;
+}
+
+void sock_dtls_session_destroy(sock_dtls_t *sock, sock_dtls_session_t *remote)
+{
+    dtls_close(sock->dtls_ctx, &remote->dtls_session);
+}
+
+ssize_t sock_dtls_send(sock_dtls_t *sock, sock_dtls_session_t *remote,
+                       const void *data, size_t len)
+{
+    assert(sock && remote && data);
+    int res;
+
+    /* check if session exists, if not create session first then send */
+    if (!dtls_get_peer(sock->dtls_ctx, &remote->dtls_session)) {
+        /* no session with remote, creating new session.
+         * This will also create new peer for this session */
+        res = dtls_connect(sock->dtls_ctx, &remote->dtls_session);
+        if (res < 0) {
+            DEBUG("sock_dtls: error initiating handshake\n");
+            return -ENOMEM;
+        }
+        else if (res > 0) {
+            /* handshake initiated, wait until connected or timed out */
+            xtimer_t timeout_timer;
+            timeout_timer.callback = _timeout_callback;
+            timeout_timer.arg = sock;
+            xtimer_set(&timeout_timer, DTLS_HANDSHAKE_TIMEOUT);
+
+            msg_t msg;
+            do {
+                mbox_get(&sock->mbox, &msg);
+            } while ((msg.type != DTLS_EVENT_CONNECTED) &&
+                     (msg.type != DTLS_EVENT_TIMEOUT));
+
+            if (msg.type == DTLS_EVENT_TIMEOUT) {
+                DEBUG("sock_dtls: handshake process timed out\n");
+
+                /* deletes peer created in dtls_connect() before */
+                dtls_peer_t *peer = dtls_get_peer(sock->dtls_ctx, &remote->dtls_session);
+                dtls_reset_peer(sock->dtls_ctx, peer);
+                return -EHOSTUNREACH;
+            }
+            xtimer_remove(&timeout_timer);
+        }
+    }
+
+    return dtls_write(sock->dtls_ctx, &remote->dtls_session, (uint8_t *)data,
+                      len);
+}
+
+ssize_t sock_dtls_recv(sock_dtls_t *sock, sock_dtls_session_t *remote,
+                       void *data, size_t max_len, uint32_t timeout)
+{
+    xtimer_t timeout_timer;
+
+    assert(sock && data && remote);
+
+    if ((timeout != SOCK_NO_TIMEOUT) && (timeout != 0)) {
+        timeout_timer.callback = _timeout_callback;
+        timeout_timer.arg = sock;
+        xtimer_set(&timeout_timer, timeout);
+    }
+
+    /* save location to result buffer */
+    sock->buf = data;
+    sock->buflen = max_len;
+
+    for (uint32_t start_recv = 0;;) {
+        start_recv = xtimer_now_usec();
+        ssize_t res = sock_udp_recv(sock->udp_sock, data, max_len, timeout,
+                                    &remote->ep);
+        if (res < 0) {
+            DEBUG("sock_dtls: error receiving UDP packet: %d\n", res);
+            return res;
+        }
+
+        if ((timeout != SOCK_NO_TIMEOUT) && (timeout != 0)) {
+            timeout = timeout - (xtimer_now_usec() - start_recv);
+            DEBUG("sock_dtls: timeout left: %"PRIu32"\n", timeout);
+        }
```

Comment:
Just to be clear: calculating the remaining timeout is just for debugging purposes, right? Maybe put a `#if ENABLE_DEBUG` around the `if` then?

Reply:
its used in L433 below again, but only compared for `0`. A possible problem here is that it rolls over due to subtraction - but I'm not sure if that is a problem at all?!

Reply:
No, calculating the remaining timeout is not just for debugging purposes. `sock_udp_recv()` is called in a loop until a DTLS packet with application data is received and each time calling it the timeout will need to be recalculated. I'll admit that with how the timeout is implemented currently, it may not be enough for applications that require high-level of timing precision.

Regarding rollover, would a check like `timeout = (timeout < 0) ? 0 : timeout` okay to prevent the rollover?

Reply:
no because `timeout` is unsigned, hence cannot be `< 0`. you need to checkout if `(xtimer_now_usec() - start_recv)` is larger or smaller than `timeout` if it is larger, set timeout to 0

Reply:
pushed a fix for this https://github.com/RIOT-OS/RIOT/pull/11943/commits/c6977c29a7de16b1693226f7eb2d96b6b7b49f0c